File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
from collections import defaultdict
from string import ascii_uppercase


def solution(msg):
    answer = []
    dic = defaultdict(int)
    n = 1
    
    for v in ascii_uppercase:
        dic[v] = n
        n += 1

    p, c = 0, 0
    
    while True:
        c += 1
        if c == len(msg):
            answer.append(dic[msg[p:]])
            return answer
        
        if msg[p:c+1] not in dic:
            dic[msg[p:c+1]] = n
            n += 1
            answer.append(dic[msg[p:c]])
            p = c

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    
    public int[] solution(String msg) {
        List<Integer> answer = new ArrayList<>();
        Map<String, Integer> dic = new HashMap<>();
        
        int n = 1;
        for (char ch = 'A'; ch <= 'Z'; ch++) {
            dic.put(String.valueOf(ch), n++);
        }
        
        int p = 0;
        int c = 0;
        
        while (true) {
            c++;
            
            if (c == msg.length()) {
                answer.add(dic.get(msg.substring(p)));
                break;
            }
            
            String next = msg.substring(p, c + 1);
            
            if (!dic.containsKey(next)) {
                dic.put(next, n++);
                answer.add(dic.get(msg.substring(p, c)));
                p = c;
            }
        }
        
        int[] result = new int[answer.size()];
        for (int i = 0; i < answer.size(); i++) {
            result[i] = answer.get(i);
        }
        
        return result;
    }
}